components/brave_sync/time_limited_words.cc

```diff
@@ -0,0 +1,172 @@
+    DCHECK(false);
+    return -1;
+  }
+
+  DCHECK_NE(mnemonic_w, nullptr);
+  size_t idx = wordlist_lookup_word(mnemonic_w, word_prepared.c_str());
+  if (!idx) {
+    return -1;
+  }
+
+  return idx - 1;
+}
+
+Time TimeLimitedWords::GetWordsV1SunsetDay() {
+  if (words_v1_sunset_day_.is_null()) {
+    bool convert_result =
+        Time::FromUTCString(kWordsv1SunsetDate, &words_v1_sunset_day_);
+    CHECK(convert_result);
+  }
+
+  CHECK(!words_v1_sunset_day_.is_null());
+
+  return words_v1_sunset_day_;
+}
+
+int TimeLimitedWords::GetRoundedDaysDiff(const Time& time1, const Time& time2) {
+  TimeDelta delta = time2 - time1;
+
+  double delta_in_days_f = delta.InMillisecondsF() / Time::kMillisecondsPerDay;
+
+  int days_rounded = std::round(delta_in_days_f);
+  return days_rounded;
+}
+
+std::string TimeLimitedWords::GenerateForNow(const std::string& pure_words) {
+  return TimeLimitedWords::GenerateForDate(pure_words, Time::Now());
+}
+
+std::string TimeLimitedWords::GenerateForDate(const std::string& pure_words,
+                                              const Time& not_after) {
+  int days_since_qrv2_epoh =
+      GetRoundedDaysDiff(QrCodeDataValidator::GetQRv1SunsetDay(), not_after);
+
+  if (days_since_qrv2_epoh < 0) {
+    // Something goes bad, requested |not_after| is even before sync v2 epoch
+    return std::string();
+  }
+
+  std::string last_word = GetWordByIndex(days_since_qrv2_epoh);
+
+  std::string time_limited_code = pure_words + " " + last_word;
+  return time_limited_code;
+}
+
+WordsValidationResult TimeLimitedWords::Validate(
+    const std::string& time_limited_words,
+    std::string* pure_words) {
+  CHECK_NE(pure_words, nullptr);
+  *pure_words = std::string();
+
+  static constexpr size_t kPureWordsCount = 24u;
+  static constexpr size_t kWordsV2Count = 25u;
+
+  auto now = Time::Now();
+
+  std::vector<std::string> words = base::SplitString(
+      time_limited_words, " ", base::WhitespaceHandling::TRIM_WHITESPACE,
+      base::SplitResult::SPLIT_WANT_NONEMPTY);
+
+  size_t num_words = words.size();
+
+  if (num_words == kPureWordsCount) {
+    if (now < GetWordsV1SunsetDay()) {
+      std::string recombined_pure_words = base::JoinString(
+          base::span<std::string>(words.begin(), kPureWordsCount), " ");
+      if (crypto::IsPassphraseValid(recombined_pure_words)) {
+        *pure_words = recombined_pure_words;
+        return WordsValidationResult::kValid;
+      } else {
+        return WordsValidationResult::kNotValidPureWords;
+      }
+    } else {
+      return WordsValidationResult::kVersionDeprecated;
+    }
+  } else if (num_words == kWordsV2Count) {
+    std::string recombined_pure_words = base::JoinString(
+        base::span<std::string>(words.begin(), kPureWordsCount), " ");
+    if (crypto::IsPassphraseValid(recombined_pure_words)) {
+      int days_actual =
+          GetRoundedDaysDiff(QrCodeDataValidator::GetQRv1SunsetDay(), now) %
```

Comment:
This should be the `epoch`, not the sunset date.

Reply:
I agree, this is confuses. I have chosen `GetQRv1SunsetDay` as an epoch for time-limited words, this works, but this is not logically correct. Thanks, I am fixing this.

Reply:
thanks, I fixed this at https://github.com/brave/brave-core/pull/12300/commits/4c59158b9070eacf74afa52c42c1930b85f942e4